### PR TITLE
[Feature&Doc]Modify ptq pipeline and support lsq

### PR DESCRIPTION
**Pull request by HIT-cwh**
## Modification

1. Quantize R18 and Retina with PTQ
2. Refactor QATLoop and support LSQLoop
3. Add LSQObserver and LSQFakequant
4. Replacing syncbn and _BatchNormXd (in mmengine) with BatchNorm2d before preparation.
5. Synchronize the training modes before fuse_fx depending on `is_qat`
6. Use special observers for ops with fixed qparams
7. Deepcopy the shared modules (e.g. shared detection head in RetinaNet) to make sure modules can be fused correctly.
8. Add the corresponding docstring.

### CODE REVIEW

File: mmrazor/models/quantizers/native_quantizer.py [hunk 1]
```diff
@@ -96,6 +103,24 @@ def __init__(self,
                 self.qconfig_mapping.set_object_type(mod, None)
         else:
             self.no_observer_modules = no_observer_modules
+
```
Comment (humu789) [inline mmrazor/models/quantizers/native_quantizer.py:106]: L98~L123 had better be a method, which can be named as `get_qconfig_mapping`

File: mmrazor/models/quantizers/academic_quantizer.py [hunk 2]
```diff
@@ -54,6 +55,9 @@ def prepare(self, model, graph_module):
             setattr(graph_module, attr_name, getattr(model, attr_name))
         fuse_custom_config = FuseCustomConfig().set_preserved_attributes(
             preserved_attributes)
+
+        self.sync_module_training_mode(graph_module, self.is_qat)
```
Comment (humu789) [inline mmrazor/models/quantizers/academic_quantizer.py:59]: 1. whether `self.sync_module_training_mode` be just used to handle some special cases? 
2. if 1 is yes, can it be use its own judgement to execute without `is_qat`
3. if 2 is no, had better not use `is_qat`, beacause `fuse` and `preprare`  do not distinguish `is_qat`, and some ptq algorithms may be to train model, such as `Adaround`

File: mmrazor/models/algorithms/quantization/mm_architecture.py [hunk 3]
```diff
@@ -54,21 +57,37 @@ def __init__(self,
...
+        # Replace syncbn and _BatchNormXd (in mmengine) with batchnorm2d
+        self.quantizer.convert_batchnorm2d(self.architecture)
+
+        if float_checkpoint:
+            _ = load_checkpoint(self.architecture, float_checkpoint)
+            self.architecture._is_init = True
 
-        self.sync_qparams('predict')
+        self.qmodels = self._build_qmodels(self.architecture)
+        self.sync_qparams('tensor')
```
Comment (humu789) [inline mmrazor/models/algorithms/quantization/mm_architecture.py:76]: had better use `predict` as default, `predict` mode is required, but `tensor` mode is optional

**Review by humu789 — COMMENTED** [review]
1. need to rebase quantize branch again
2. need to pass ci

File: mmrazor/models/fake_quants/lsq.py [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+    """Enable validation, if applicable. Example usage::
+
+    # model is any PyTorch model model.apply(enable_val)
+    """
+    if isinstance(mod, LearnableFakeQuantize):
+        mod.enable_val()
+
+
+@MODELS.register_module()
+class LearnableFakeQuantize(FakeQuantizeBase):
```
[the remark this answers is not in the document]
Comment (HIT-cwh) [reply mmrazor/models/fake_quants/lsq.py:43]: `_LearnableFakeQuantize` in Pytorch inherits from FakeQuantizeBase. And many interfaces are different between `_LearnableFakeQuantize` and `FakeQuantize` (`BaseFakeQuantize ` in razor) such as `calculate_qparams`.

File: mmrazor/models/algorithms/quantization/mm_architecture.py (hunk 3 above)
Comment (HIT-cwh) [reply mmrazor/models/algorithms/quantization/mm_architecture.py:76]: `data_samples` can not be `None` in detectors during prediction. But we need to make the dummy prediction in `_build_qmodels`. It is more convenient to use tensor mode.

File: mmrazor/models/quantizers/academic_quantizer.py (hunk 2 above)
Comment (HIT-cwh) [reply mmrazor/models/quantizers/academic_quantizer.py:59]: Successive conv bn layers may be in different modes. For example, bn layers in detectors are usually in eval mode during training. But the successive conv bn layers should be in the same mode during `_fuse_fx`. 
It seems that all modules can be set to training mode. I fixed it just now.

**Review by humu789 — APPROVED** [review]
